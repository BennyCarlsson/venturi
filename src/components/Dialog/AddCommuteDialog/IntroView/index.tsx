import Button from 'components/Button'
import Heading from 'components/Heading'
import image from 'assets/images/bus-stop/background-bus-stop.jpg'
import {
  BottomPart,
  IntroViewWrapper,
  StyledImage,
  StyledParagraph,
  StyledTitleHeading,
} from './styles'

type IntroViewProps = {
  goToChooseDestionationView: () => void
}

const IntroView = ({ goToChooseDestionationView }: IntroViewProps) => {
  return (
    <IntroViewWrapper>
      <span>
        <StyledTitleHeading
          headingType={'h1'}
          fontSize={28}
          color='#719FB0'
          fontWeight={400}
        >
          Venturi
        </StyledTitleHeading>
        <Heading
          headingType='h2'
          fontSize={28}
          fontWeight={700}
          color={'#35F39'}
        >
          Commute trip planner
        </Heading>
        <StyledParagraph fontSize={14} fontWeight={500}>
          Have a quick way to get information about your most common trip.
        </StyledParagraph>
        <StyledParagraph fontSize={14} fontWeight={500}>
          The specified trip will be easily acceible with current status and
          more.
        </StyledParagraph>
      </span>
      <BottomPart>
        <StyledImage src={image} />
        <Button onClick={goToChooseDestionationView}>Get Started!</Button>
      </BottomPart>
    </IntroViewWrapper>
  )
}

export default IntroView
