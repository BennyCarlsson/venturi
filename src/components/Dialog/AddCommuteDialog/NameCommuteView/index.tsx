import Paragraph from 'components/Paragraph'
import React from 'react'
import {
  CTAButtonTitleWrapper,
  Input,
  StyledButton,
  StyledClosedIconButton,
  StyledParagraph,
  StyledTitleHeading,
  TopCTAButtonTitleSubtitle,
} from './styles'

type NameCommuteViewProps = {
  createTrip: () => void
  goToChooseDestionationView: () => void
}

const NameCommuteView = ({
  createTrip,
  goToChooseDestionationView,
  ...props
}: NameCommuteViewProps) => {
  return (
    <React.Fragment {...props}>
      <TopCTAButtonTitleSubtitle>
        <CTAButtonTitleWrapper>
          <StyledClosedIconButton
            data-testid='close-dialog-button'
            onClick={(e) => {
              goToChooseDestionationView()
            }}
          />
          <StyledTitleHeading headingType={'h2'} fontSize={28}>
            Your Trip
          </StyledTitleHeading>
        </CTAButtonTitleWrapper>
        <Paragraph fontSize={14}>
          To kepp track of your trips, a good idea is to name the specific trip
          you are about to create. It's optional but a good idea.
        </Paragraph>
        <StyledParagraph fontSize={18}>Trip name</StyledParagraph>
        <Input />
      </TopCTAButtonTitleSubtitle>
      <StyledButton onClick={createTrip}>Create Commute trip</StyledButton>
    </React.Fragment>
  )
}

export default NameCommuteView
