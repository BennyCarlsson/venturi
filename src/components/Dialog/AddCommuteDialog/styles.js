import styled from 'styled-components';

export const AddCommuteDialogWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 100%;
  padding: 16px;
`;
