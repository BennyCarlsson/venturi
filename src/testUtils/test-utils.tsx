import { configureStore } from '@reduxjs/toolkit';
import locationsReducer from '../redux/locationsSlice';
import testReducer from '../redux/testSlice';
import tripReducer from '../redux/tripSlice';
import dialogReducer from '../redux/dialogSlice';
import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ReactElement } from 'react';
import { RootState } from 'redux/store';

type RenderOptionsType = { initialState?: Partial<RootState> } & Omit<RenderOptions, 'queries'>;
function render(ui: ReactElement, { initialState = {}, ...renderOptions }: RenderOptionsType) {
  const store = configureStore({
    reducer: {
      test: testReducer,
      trip: tripReducer,
      locations: locationsReducer,
      dialog: dialogReducer
    },
    preloadedState: initialState
  });
  function Wrapper({ children }: any) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { render };
