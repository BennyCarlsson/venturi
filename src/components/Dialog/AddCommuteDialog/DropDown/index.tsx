import { Fragment } from 'react';
import { Location } from 'redux/locationsSlice';
import { Li, Ul } from './styles';

type DropDownProps = {
  locations: Location[] | undefined;
  setInput: (value: React.SetStateAction<string>) => void;
  setLocation: (value: React.SetStateAction<Location | null>) => void;
  show: boolean;
};

const DropDown = ({ locations, setInput, setLocation, show }: DropDownProps) => {
  const shouldRender = () => show && Array.isArray(locations) && locations.length > 0;
  return (
    <Fragment>
      {shouldRender() && (
        <Ul>
          {locations?.map((loc) => (
            <Li
              key={`key-${loc.id}`}
              onClick={(e) => {
                e.preventDefault();
                setInput(loc.name);
                setLocation(loc);
              }}
            >
              {loc.name}
            </Li>
          ))}
        </Ul>
      )}
    </Fragment>
  );
};

export default DropDown;
