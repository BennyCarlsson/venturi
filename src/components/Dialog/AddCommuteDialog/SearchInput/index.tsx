import { Dispatch, SetStateAction } from "react";
import { Input } from "./styles";

type SearchInputProps = {
  value: string;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  setShowDropDown: Dispatch<SetStateAction<boolean>>;
};

const SearchInput = ({
  value,
  onChange,
  setShowDropDown,
  ...rest
}: SearchInputProps) => {
  return (
    <Input
      {...rest}
      value={value}
      onChange={onChange}
      onBlur={() => {
        setTimeout(() => {
          setShowDropDown(false);
        }, 100);
      }}
      onFocus={() => {
        if (value.length > 1) {
          setShowDropDown(true);
        }
      }}
    />
  );
};

export default SearchInput;
