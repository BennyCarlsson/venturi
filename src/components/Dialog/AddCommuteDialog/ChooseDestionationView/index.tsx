import Paragraph from 'components/Paragraph';
import { useAppDispatch, useAppSelector } from 'hooks/redux';
import React, { useCallback, useRef, useState } from 'react';
import {
  fetchDestinationLocationsOnName,
  fetchOriginLocationsOnName,
  Location
} from 'redux/locationsSlice';
import { setCurrentTable } from 'redux/tripSlice';
import DropDown from '../DropDown';
import SearchInput from '../SearchInput';
import { saveOriginAndDestination } from 'utils/localStorage';
import { StyledButton, StyledParagraph, StyledTitleHeading, StyledToParagraph } from './styles';

let timeout: any;
const debounce = (fn: Function, delay: number) => {
  if (timeout !== undefined) {
    clearTimeout(timeout);
  }
  timeout = setTimeout(fn, delay);
};

type ChooseDestionationViewProps = {};

const ChooseDestionationView = ({ ...props }: ChooseDestionationViewProps) => {
  const [originInput, setOriginInput] = useState('');
  const [showOriginDropDown, setShowOriginDropDown] = useState(false);
  const [destinationInput, setDestinationInput] = useState('');
  const [showDestinationDropDown, setShowDestinationDropDown] = useState(false);
  const [originLocation, setOriginLocation] = useState<Location | null>(null);
  const [destinationLocation, setDestinationLocation] = useState<Location | null>(null);

  const dispatch = useAppDispatch();
  const originLocations = useAppSelector((state) => state.locations.originLocations);
  const destinationLocations = useAppSelector((state) => state.locations.destinationLocations);
  const loading = useAppSelector((state) => state.locations.loadingDestination);
  const error = useAppSelector((state) => state.locations.destinationError);
  let promise = useRef<any>();

  const originInputOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setOriginInput(value);
    debounce(() => {
      requestOrigin(value);
    }, 500);
  };

  const requestOrigin = useCallback(
    (value: string) => {
      if (value.length > 1) {
        if (promise.current && !error && loading) {
          promise.current.abort();
        }
        setShowOriginDropDown(true);
        promise.current = dispatch(fetchOriginLocationsOnName(value));
      } else {
        setShowOriginDropDown(false);
      }
    },
    [dispatch, error, loading]
  );

  const destinationOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setDestinationInput(value);
    debounce(() => {
      requestDestination(value);
    }, 500);
  };
  const requestDestination = useCallback(
    (value: string) => {
      if (value.length > 1) {
        if (promise.current && !error && loading) {
          promise.current.abort();
        }
        setShowDestinationDropDown(true);
        promise.current = dispatch(fetchDestinationLocationsOnName(value));
      } else {
        setShowDestinationDropDown(false);
      }
    },
    [dispatch, error, loading]
  );
  return (
    <React.Fragment {...props}>
      <span>
        <StyledTitleHeading headingType={'h2'} fontSize={28}>
          Destination
        </StyledTitleHeading>
        <StyledParagraph fontSize={14}>
          Välj destinationer för din resa. Du kan alltid ändra eller lägga till flera vid ett senare
          tillfälle.
        </StyledParagraph>
        <Paragraph fontSize={18}>Från</Paragraph>
        <SearchInput
          value={originInput}
          onChange={originInputOnChange}
          setShowDropDown={setShowOriginDropDown}
        />
        <DropDown
          locations={originLocations}
          setInput={setOriginInput}
          setLocation={setOriginLocation}
          show={showOriginDropDown}
        />
        <StyledToParagraph fontSize={18}>Till</StyledToParagraph>
        <SearchInput
          value={destinationInput}
          onChange={destinationOnChange}
          setShowDropDown={setShowDestinationDropDown}
        />
        <DropDown
          locations={destinationLocations}
          setInput={setDestinationInput}
          setLocation={setDestinationLocation}
          show={showDestinationDropDown}
        />
      </span>
      <StyledButton
        onClick={() => {
          if (originLocation && destinationLocation)
            dispatch(setCurrentTable({ origin: originLocation, destination: destinationLocation }));
          saveOriginAndDestination({
            origin: originLocation as Location,
            destination: destinationLocation as Location
          });
        }}
      >
        Namnge din resa
      </StyledButton>
      {error && <h2>{error}</h2>}
    </React.Fragment>
  );
};

export default ChooseDestionationView;
