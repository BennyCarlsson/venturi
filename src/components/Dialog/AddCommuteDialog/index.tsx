import { useState } from 'react'
import ChooseDestionationView from './ChooseDestionationView'
import NameCommuteView from './NameCommuteView'
import { AddCommuteDialogWrapper } from './styles'
import { Location } from 'redux/locationsSlice'
import { useAppDispatch } from 'hooks/redux'
import { setCurrentTable } from 'redux/tripSlice'
import { saveOriginAndDestination } from 'utils/localStorage'
import { hideDialog } from 'redux/dialogSlice'
import { addTripToList } from 'redux/tripListSlice'
import IntroView from './IntroView'

type AddCommuteDialogProps = {
  isIntroView: boolean
}

const AddCommuteDialog = ({ isIntroView, ...props }: AddCommuteDialogProps) => {
  const dispatch = useAppDispatch()
  const [displayView, setDisplayView] = useState<
    'IntroView' | 'ChooseDestinationView' | 'NameCommuteView'
  >(isIntroView ? 'IntroView' : 'ChooseDestinationView')
  const [originLocation, setOriginLocation] = useState<Location | null>(null)
  const [name, setName] = useState('')
  const [destinationLocation, setDestinationLocation] =
    useState<Location | null>(null)

  const goToIntroView = () => {
    setDisplayView('IntroView')
  }

  const goToNameCommuteView = () => {
    setDisplayView('NameCommuteView')
  }

  const goToChooseDestinationView = () => {
    setDisplayView('ChooseDestinationView')
  }

  const createTrip = () => {
    if (originLocation && destinationLocation) {
      dispatch(
        setCurrentTable({
          name,
          origin: originLocation,
          destination: destinationLocation,
        })
      )
      saveOriginAndDestination({
        currentTripName: name,
        origin: originLocation,
        destination: destinationLocation,
      })
      dispatch(
        addTripToList({
          name: name,
          origin: originLocation,
          destination: destinationLocation,
        })
      )
      dispatch(hideDialog())
    }
  }
  return (
    <AddCommuteDialogWrapper {...props}>
      {displayView === 'IntroView' && (
        <IntroView goToChooseDestionationView={goToChooseDestinationView} />
      )}
      {displayView === 'ChooseDestinationView' && (
        <ChooseDestionationView
          goBack={() => {
            isIntroView ? goToIntroView() : dispatch(hideDialog())
          }}
          goToNameCommuteView={goToNameCommuteView}
          originLocation={originLocation}
          setOriginLocation={setOriginLocation}
          destinationLocation={destinationLocation}
          setDestinationLocation={setDestinationLocation}
        />
      )}
      {displayView === 'NameCommuteView' && (
        <NameCommuteView
          name={name}
          setName={setName}
          createTrip={createTrip}
          goToChooseDestionationView={goToChooseDestinationView}
        />
      )}
    </AddCommuteDialogWrapper>
  )
}

export default AddCommuteDialog
