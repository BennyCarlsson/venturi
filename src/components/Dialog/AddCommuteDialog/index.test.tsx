import AddCommuteDialog from '.';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { useDispatch } from 'react-redux';
import { cleanup, render, screen } from 'testUtils/test-utils';

jest.mock('react-redux', () => ({
  ...(jest.requireActual('react-redux') as {}),
  useDispatch: jest.fn()
}));

describe('<AddCommuteDialog />', () => {
  const useAppDispatch = jest.fn();
  beforeEach(() => {
    (useDispatch as jest.Mock).mockReturnValue(useAppDispatch);
  });

  it('should match snapshot', () => {
    cleanup();
    const { asFragment } = render(<AddCommuteDialog />, {});
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render without crashing', () => {
    render(<AddCommuteDialog />, {
      initialState: {
        dialog: { show: true },
        locations: {
          destinationError: 'error',
          loadingDestination: false,
          loadingOrigin: false
        }
      }
    });
    expect(screen).toBeTruthy();
  });
});
