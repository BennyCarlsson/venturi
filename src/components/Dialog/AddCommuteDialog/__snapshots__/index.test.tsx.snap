// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCommuteDialog /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-jrAGrp hffLVH"
  >
    <h2
      class="sc-jSgupP ehQrep sc-iqHYGH bVotdH"
      font-size="28"
    >
      Destination
    </h2>
    <p
      class="sc-bdfBwQ QuMsF sc-crrsfI dJBLtM"
      font-size="14"
    >
      Choose your commute trip
    </p>
    <p
      class="sc-bdfBwQ kMVXxV"
      font-size="18"
    >
      From
    </p>
    <input
      class="sc-hKgILt jroXQp"
      value=""
    />
    <p
      class="sc-bdfBwQ kMVXxV sc-kEjbxe hVozjA"
      font-size="18"
    >
      To
    </p>
    <input
      class="sc-hKgILt jroXQp"
      value=""
    />
    <button>
      Save
    </button>
  </div>
</DocumentFragment>
`;
